### DISCUSSION TITLE
Adding lever arm to generated gnss data

### DISCUSSION THREAD

**Discussion by Mostafa-NA** [rodralez/NaveGo#100] [Q&A] [answered]
Dear team of NaveGo,
Thank you for the effort and dedication put in this code.
i wanted to test the effect of accounting the lever arm or dismissing it, so i modified the gnss_gen.m function to add a lever arm of [0.2 , -0.3 ,0] (in meters) to the generated GNSS data (in LLH). i did that by adding and modifying the gnss_gen.m as follows

%%%%%%%%%%%%%%%%%%%%%%%%%
Lev_arm = zeros(size(gnss.t),3);

for i=1:size(gnss.t)

[RM,RN] = radius(gnss_r.lat(i));
Tpr = diag([(RM + gnss_r.h(i)), (RN + gnss_r.h(i)) * cos(gnss_r.lat(i)), -1]);
Trp = inv(Tpr);
DCMbn = inv([ref.DCMnb(i,1:3);ref.DCMnb(i,4:6);ref.DCMnb(i,7:9)]);
Lev_arm(i,:) = Trp*DCMbn*gnss.larm; % gnss.larm = [0.2 , -0.3 ,0] was defined in navego_example_synth.m 

end

gnss.t   = gnss_r.t;
gnss.lat = gnss_r.lat + gnss.std(1) .* r1 + Lev_arm(:,1);
gnss.lon = gnss_r.lon + gnss.std(2) .* r2 + Lev_arm(:,2);
gnss.h   = gnss_r.h   + gnss.std(3) .* r3 + Lev_arm(:,3);
gnss.vel = gnss_r.vel + [gnss.stdv(1).*r4  gnss.stdv(2).*r5  gnss.stdv(3).*r6];

gnss.larm = zeros(3,1);  % comment this line to account for the lever arm, else the code will not correct for it

end
%%%%%%%%%%%%%%%%%%%%%%%%%%

is this method valid or not? and doesn't the lever arm also affect the gnss velocity?
please correct me and point out some references regarding this topic. If you consider adding this feature to your code, this would be fantastic. 
Thank you again

**Comment by rodralez** [comment, marked as answer]
Hello,

Thank you for your kind words.

It seems that your code is not right.

In function `ins_gnss.m` lever arm effect is corrected for position (lines [330](https://github.com/rodralez/NaveGo/blob/00aaba52c28225d119ff1db98b4af6f752d573ef/ins-gnss/ins_gnss.m#L330) and 331), and for velocities (lines [334](https://github.com/rodralez/NaveGo/blob/00aaba52c28225d119ff1db98b4af6f752d573ef/ins-gnss/ins_gnss.m#L334) and 335).

Your code has to do the opposite of what these two formulas do:

1. Substracting `(DCMbn * gnss.larm)` for position.
2. Adding  `((omega_ie_n + omega_en_n) * (DCMbn * gnss.larm))' - (DCMbn * skewm(wb_corrected) * gnss.larm)' )` for velocities.

Kind regards.

**Reply by Mostafa-NA** [reply]
Thanks for your reply.
I understand now the idea, but i still have a question, shouldn't i multiply (DCMbn * gnss.larm) with the inverse of Tpr before subtracting it, since in the correction of the lever arm effect in ins_gnss.m (line 330 and 331) we multiply the gnss components by Tpr and then adds (DCMbn * gnss.larm)?

**Reply by rodralez** [reply]
The Tpr matrix transforms radians to meters for a curvilinear trajectory over the Earth surface. To do the opposite, please check Eq. 2.119 in Groves' book.
